Add documentation - Main Concepts - Data Storage (#1860)
kuzzleio/kuzzle: doc/2/guides/main-concepts/2-data-storage/index.md

This action takes as parameter a definition of a set of indexes and collections with their associated mappings.  

The expected format is the following:
```js
{
  "<index>": {
    "<collection>": {
      "properties": {
        "<field>": {}
      }
    }
  },
  "<index>": {
    "<collection>": {
      "properties": {
        "<field>": {}
      }
    }
  }
}

```

::: info
It is recommended to **save the mappings of your application in a JSON file**.
:::

```bash
# Create a file containing the mappings
echo '
{
  "nyc-open-data": {
    "yellow-tax": {
      "properties": {
        "name": { "type": "keyword" },
        "age": { "type": "integer" }
      }
    }
  }
}
' > mappings.json

# Load it with Kourou
kourou admin:loadMappings < mappings.json
```

::: info
Kourou can read file content and put it the request body.
:::

<!-- 
  @todo load at startup with Kaaf
-->

## Collection Settings

In addition mappings, Elasticsearch exposes [settings](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/index-modules.html#index-modules-settings) that allow to finely configure the behavior of a collection.

> Those settings allows to configure [custom analyzers](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/analysis-custom-analyzer.html) for example.

Reviewer: ```suggestion
These settings allow to configure [custom analyzers](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/analysis-custom-analyzer.html) for example.
```

```bash
kourou collection:create ktm-open-data:yellow-taxi '{
  "settings": {
    "analysis": {
      "analyzer": {
        "my_custom_analyzer": {
          "type": "custom", 
          "tokenizer": "standard",
          "char_filter": [
            "html_strip"
          ],
          "filter": [
            "lowercase",
            "asciifolding"
          ]
        }
      }
    }
  }
}'
```

::: warning
While updating the collection settings, the collection will be [closed](https://www.elastic.co/guide/en/elasticsearch/reference/7.5/indices-close.html) until the new configuration has been applied.
:::

## Kuzzle Metadata

> Whenever a **document is created, updated or deleted**, Kuzzle will **add or update the document's metadata**. Those metadata provides information about the document's lifecycle.

Reviewer: ```suggestion
Whenever a **document is created, updated or deleted**, Kuzzle will **add or update the document's metadata**. These metadata provide information about the document's lifecycle.
```

:::

These actions work in the same way as single document actions (`createOrReplace` becomes `mCreateOrReplace`, `update` becomes `mUpdate`, etc) but by taking an array of documents in the `body` request.

**Example:** _Create multiple documents with the [document:mCreate](/core/2/api/controllers/document/m-create) action_

<!-- 
  @todo deprecate "body" and use "content" instead
-->

```bash
kourou document:mCreate ktm-open-data thamel-taxi '{
  documents: [
    {
      _id: "liia-mery",
      body: {
        driver: {
          name: "liia mery"
        },
        category: "limousine"
      }
    },
    {
      body: {
        driver: {
          name: "aschen"
        },
        category: "suv"
      }
    }
  ]
}'
```

### Write Limit

Kuzzle imposes a **limit to the number of documents that can be written by the same request**.

This limit ensures that Kuzzle and Elasticsearch are not overloaded by writing too many documents at once.

By default, this limit is `200` documents per request. It is possible to configure this value in the `limits.documentsWriteCount` key in the [configuration](/core/2/guides/advanced/8-configuration) file.

## Read Documents

> Kuzzle exposes methods to read documents. There is two way of retrieving documents:

Reviewer: ```suggestion
Kuzzle exposes methods to read documents. There are two way of retrieving documents:
```

>  - by document `_id`

Reviewer: ```suggestion
 - using a document unique identifier (`_id` field)
```

This limit ensures that Kuzzle and Elasticsearch are not overloaded by returning too many documents at once.

By default, this limit is `10000` documents per request. It is possible to configure this value in the `limits.documentsFetchCount` key in the [configuration](/core/2/guides/advanced/8-configuration) file.

## Bulk Actions

To perform lower level actions, it is possible to use the [bulk](/core/2/api/controllers/bulk) controller.

The actions of this controller may not follow some of the API principles such as:
 - adding [Kuzzle Metadata](/core/2/guides/main-concepts/2-data-storage#kuzzle-metadata)
 - triggering [Database Notifications](/core/2/guides/main-concepts/6-realtime-engine#database-notifications) 
 - application of [Data Validation](/core/2/guides/advanced/9-data-validation) rules
 - respect of [write limit](/core/2/guides/main-concepts/2-data-storage#some-anchor)

The following actions are available:
 - [bulk:write](/core/2/api/controllers/bulk/write): write a document
 - [bulk:mWrite](/core/2/api/controllers/bulk/m-write): write multiple documents
 - [bulk:import](/core/2/api/controllers/bulk/import): import documents as fast as possible

>  - [bulk:deleteByQuery](/core/2/api/controllers/bulk/write): deletes large volume of documents matching a query

Reviewer: ```suggestion
 - [bulk:deleteByQuery](/core/2/api/controllers/bulk/write): delete large volume of documents matching a query
```

::: warning

> Bulk actions are intended to be used by administrator and scripts.  

Reviewer: ```suggestion
Bulk actions are intended to be used by administrators and scripts.  
```

It is considered harmful to let end users execute those actions.
:::

## Integrated Elasticsearch Client

> Kuzzle uses and exposes [Elasticsearch Javascript SDK](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html). 

Reviewer: ```suggestion
Kuzzle uses and exposes the [Elasticsearch Javascript SDK](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html). 
```

It is possible to **interact directly with Elasticsearch** through clients exposed in the [Backend.storage](/core/2/some-link) property.

This property offers the possibility to **instantiate a new client** or to **use a lazy-instantiated client**. In both cases, the clients are configured to use the same Elasticsearch cluster as Kuzzle.

::: info

> It is possible to overload the configuration used by default by instantiating a new Ealsticsearch client with the constructor [Backend.storage.Client](/core/2/some-link).

Reviewer: ```suggestion
It is possible to overload the configuration used by default by instantiating a new Elasticsearch client with the constructor [Backend.storage.Client](/core/2/some-link).
```

:::

**Example:** _Send a request directly to Elasticsearch_

```js
// Elasticsearch request to create a document
const esRequest =  {
  body: {
    name: 'Aschen',
    age: 27
  },
  // Internal name of the index "nyc-open-data" and the collection "yellow-taxi"
  index: '%nyc-open-data.yellow-taxi',
  op_type: 'create'
}

// Use directly an Elasticsearch client instance
await app.storage.client.index(esRequest)

// Instantiate and use a new client
const esClient = new app.storage.Client()
await esClient.index(esRequest)
```

::: warning

> Kuzzle use an [internal naming system](/core/2/guides/main-concepts/2-data-storage#some-anchor) to map Elasticsearch index names with Kuzzle indexes and collections names.

Reviewer: ```suggestion
Kuzzle uses an [internal naming system](/core/2/guides/main-concepts/2-data-storage#some-anchor) to map Elasticsearch index names with Kuzzle indexes and collections names.
```
